nrf-hal-common/src/comp.rs

```diff
@@ -0,0 +1,281 @@
+//! HAL interface for the COMP peripheral.
+//!
+//! The comparator (COMP) compares an input voltage (Vin) against a second input voltage (Vref).
+//! Vin can be derived from an analog input pin (AIN0-AIN7).
+//! Vref can be derived from multiple sources depending on the operation mode of the comparator.
+
+use {
+    crate::gpio::{Floating, Input},
+    crate::pac::comp::{_EVENTS_CROSS, _EVENTS_DOWN, _EVENTS_UP},
+    crate::pac::{generic::Reg, COMP},
+};
+
+/// A safe wrapper around the `COMP` peripheral.
+pub struct Comp {
+    comp: COMP,
+}
+
+impl Comp {
+    /// Takes ownership of the `COMP` peripheral, returning a safe wrapper.
+    pub fn new<P: AnalogPin>(comp: COMP, input_pin: &P) -> Self {
+        comp.psel.write(|w| match input_pin.ain() {
+            0 => w.psel().analog_input0(),
+            1 => w.psel().analog_input1(),
+            2 => w.psel().analog_input2(),
+            3 => w.psel().analog_input3(),
+            4 => w.psel().analog_input4(),
+            5 => w.psel().analog_input5(),
+            6 => w.psel().analog_input6(),
+            #[cfg(not(feature = "52810"))]
+            7 => w.psel().analog_input7(),
+            _ => unreachable!(),
+        });
+        comp.mode.write(|w| w.sp().normal());
+        comp.mode.write(|w| w.main().se());
+        comp.refsel.write(|w| w.refsel().int1v2());
+        Self { comp }
+    }
+
+    /// Sets the speed and power mode of the comparator.
+    #[inline(always)]
+    pub fn power_mode(&self, mode: PowerMode) -> &Self {
+        match mode {
+            PowerMode::LowPower => self.comp.mode.write(|w| w.sp().low()),
+            PowerMode::Normal => self.comp.mode.write(|w| w.sp().normal()),
+            PowerMode::HighSpeed => self.comp.mode.write(|w| w.sp().high()),
+        }
+        self
+    }
+
+    /// Sets Vref of the comparator in single ended mode.
+    #[inline(always)]
+    pub fn vref(&self, vref: VRef) -> &Self {
+        self.comp.refsel.write(|w| match vref {
+            VRef::Int1V2 => w.refsel().int1v2(),
+            VRef::Int1V8 => w.refsel().int1v8(),
+            VRef::Int2V4 => w.refsel().int2v4(),
+            VRef::Vdd => w.refsel().vdd(),
+            VRef::ARef => w.refsel().aref(),
+        });
+        self
+    }
+
+    /// Sets analog reference pin.
+    #[inline(always)]
+    pub fn aref_pin<P: AnalogPin>(&self, ref_pin: &P) -> &Self {
+        self.comp.extrefsel.write(|w| match ref_pin.ain() {
+            0 => w.extrefsel().analog_reference0(),
+            1 => w.extrefsel().analog_reference1(),
+            2 => w.extrefsel().analog_reference2(),
+            3 => w.extrefsel().analog_reference3(),
+            4 => w.extrefsel().analog_reference4(),
+            5 => w.extrefsel().analog_reference5(),
+            6 => w.extrefsel().analog_reference6(),
+            7 => w.extrefsel().analog_reference7(),
+            _ => unreachable!(),
```

Comment:
Maybe use an enum instead of an index for this? You can probably return an `EXTREFSEL_A` from your `ain()` function and just use do `w.extrefsel().variant(ref_pin.ain())`.

 https://docs.rs/nrf52840-pac/0.9.0/nrf52840_pac/comp/extrefsel/enum.EXTREFSEL_A.html

Reply:
Good idea! Did the same with PSEL_A for input